### DISCUSSION TITLE
hub/inference-api/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#13055] [Docs]
# hub/inference-api/

Effortlessly run AI model inferences with Ultralytics HUB Inference API. Perfect for developers!.

https://docs.ultralytics.com/hub/inference-api/

**Comment by Hamzaboughfir123** [comment]
how can I do that I have already my model trained so how can I use ultralytics hub to deploy

**Reply by pderrenger** [reply]
Hello! Great to hear that you have your model trained and are ready to deploy it. 🚀

You can easily use the Ultralytics HUB Inference API to deploy your trained model. Here’s a quick guide:

1. **Get your API Key and Model ID:** You'll need these to authenticate and specify which model to use when making API requests.
2. **Prepare your API request:** You can use either Python or cURL to send your requests. Here’s a brief example using Python:

```python
import requests

# Replace 'MODEL_ID' with your model's ID and 'API_KEY' with your actual API key
url = "https://api.ultralytics.com/v1/predict/MODEL_ID"
headers = {"x-api-key": "API_KEY"}
data = {"size": 640, "confidence": 0.25, "iou": 0.45}

# Replace 'path/to/image.jpg' with the path to your image file
with open("path/to/image.jpg", "rb") as image_file:
    files = {"image": image_file}
    response = requests.post(url, headers=headers, files=files, data=data)

print(response.json())
```

3. **Send the request:** Use the script to send a request to the API. The API will process the image using your model and return the inference results.

This setup allows you to deploy your model without setting up a local server, making it straightforward and efficient to start making predictions right away!

If you need more detailed instructions or have any other questions, feel free to ask. Happy deploying! 🌟

**Comment by SumitJana404** [comment]
Is it possible to perform inference on multiple images in a single api call?
I have tried but while printing the request.json(), its only giving inference data of a single image

import requests
url = "https://api.ultralytics.com/v1/predict/Model_id"
headers = {"x-api-key": "API_KEY"}
data = {"size": 640, "confidence": 0.25, "iou": 0.45}

image_paths = ["/content/106999_XL1.jpg", "/content/219982_XL1.jpg", "/content/258071_LXL1_20210222.jpg"]
files = [("image", open(image_path, "rb")) for image_path in image_paths]

response = requests.post(url, headers=headers, files=files, data=data)

print(response.json())
for _, file in files:
    file.close()


#Output:
{'data': [{'class': 0, 'confidence': 0.9573774337768555, 'height': 0.9333934187889099, 'name': 'object', 'width': 0.5947443246841431, 'xcenter': 0.5113524198532104, 'ycenter': 0.5011465549468994}], 'message': 'Inference complete.', 'success': True}

**Reply by pderrenger** [reply]
@SumitJana404 hi there!

Thank you for reaching out. Currently, the Ultralytics HUB Inference API processes one image per request. This is why you're seeing inference data for only a single image in your response.

To perform inference on multiple images, you would need to make separate API calls for each image. Here’s a modified version of your code to handle multiple images sequentially:

```python
import requests

# API URL, use actual MODEL_ID
url = "https://api.ultralytics.com/v1/predict/MODEL_ID"

# Headers, use actual API_KEY
headers = {"x-api-key": "API_KEY"}

# Inference arguments (optional)
data = {"size": 640, "confidence": 0.25, "iou": 0.45}

image_paths = ["/content/106999_XL1.jpg", "/content/219982_XL1.jpg", "/content/258071_LXL1_20210222.jpg"]

for image_path in image_paths:
    with open(image_path, "rb") as image_file:
        files = {"image": image_file}
        response = requests.post(url, headers=headers, files=files, data=data)
        print(response.json())
```

This script will iterate over your list of image paths, sending each image in a separate request and printing the response for each.

For more details on using the Ultralytics HUB Inference API, you can refer to our [documentation](https://docs.ultralytics.com/hub/inference-api/).

If you have any further questions or need additional assistance, feel free to ask! 😊

**Comment by smlngspprt** [comment]
What does `width`, `height` , `xcenter` and `ycenter` represent in json response? ANe what is the unit of of `wdith` and `height`
I wanted to draw boxes in the image based on inference result. What would be the best way to achieve that?

**Reply by glenn-jocher** [reply]
@smlngspprt the `width`, `height`, `xcenter`, and `ycenter` in the JSON response represent the dimensions and center coordinates of the bounding box around the detected object. These values are normalized between 0 and 1, relative to the image dimensions. To draw boxes on the image, you can convert these normalized values back to pixel coordinates based on the image's width and height. For more details on using the API, please visit [Ultralytics HUB Inference API](https://docs.ultralytics.com/hub/inference-api/).

**Comment by S-Y-3-D** [comment]
Hi, how can I send a video file to the Ultralytics Inference API? Can you please provide an example. Thanks.

**Reply by glenn-jocher** [reply]
@S-Y-3-D to send a video file to the Ultralytics Inference API, you can use a similar approach as with images. Here's a brief example using Python:

```python
import requests

# API URL, use actual MODEL_ID
url = "https://api.ultralytics.com/v1/predict/MODEL_ID"

# Headers, use actual API_KEY
headers = {"x-api-key": "API_KEY"}

# Inference arguments (optional)
data = {"imgsz": 640, "conf": 0.25, "iou": 0.45}

# Load video and send request
with open("path/to/video.mp4", "rb") as video_file:
    files = {"file": video_file}
    response = requests.post(url, headers=headers, files=files, data=data)

print(response.json())
```

Replace `MODEL_ID`, `API_KEY`, and `path/to/video.mp4` with your actual values. For more details, visit [Ultralytics HUB Inference API documentation](https://docs.ultralytics.com/hub/inference-api/).

**Comment by LyWilson** [comment]
Hi,

I'm using the free version of Ultralytics Hub and is excellent!

But currently I'm facing a issues, is that I train something that wasn't suppose to be trained, so I'm wondering if I can like disable the detection of the class object. because sometime it get confuse and make a wrong detection.

**Reply by glenn-jocher** [reply]
Thank you for your kind words about Ultralytics HUB! To disable the detection of a specific class, you can adjust the confidence threshold during inference to a higher value, effectively ignoring lower-confidence detections. You can do this by modifying the `conf` parameter in your API call. If you need more detailed guidance, consider checking our [Ultralytics HUB documentation](https://docs.ultralytics.com/hub/inference-api/).
